File: tests/bench_algorithms.py
"""Benchmarks for string distance/similarity algorithms (pytest-benchmark)."""

import itertools

import pytest

from kaos_nlp_core.algorithms import (
    damerau_levenshtein,
    hamming,
    jaro,
    jaro_winkler,
    lcs_distance,
    lcs_length,
    levenshtein,
    longest_common_substring_length,
    metaphone_encode,
    ngram_cosine,
    ngram_jaccard,
    osa,
    sorensen_dice,
    soundex_encode,
    token_jaccard,
    token_ngram_cosine,
    token_ngram_jaccard,
    token_ngram_overlap,
)

SHORT_PAIRS = [
    ("kitten", "sitting"),
    ("martha", "marhta"),
    ("Robert", "Rupert"),
    ("William", "Williams"),
    ("Philadelphia", "Philadlephia"),
    ("algorithm", "altruistic"),
]


# --- Edit distance: short strings ---


@pytest.mark.benchmark(group="edit_distance_short")
def test_levenshtein_short(benchmark):
    benchmark(lambda: [levenshtein(a, b) for a, b in SHORT_PAIRS])


@pytest.mark.benchmark(group="edit_distance_short")
def test_damerau_levenshtein_short(benchmark):
    benchmark(lambda: [damerau_levenshtein(a, b) for a, b in SHORT_PAIRS])


@pytest.mark.benchmark(group="edit_distance_short")
def test_osa_short(benchmark):
    benchmark(lambda: [osa(a, b) for a, b in SHORT_PAIRS])


@pytest.mark.benchmark(group="edit_distance_short")
def test_jaro_short(benchmark):
    benchmark(lambda: [jaro(a, b) for a, b in SHORT_PAIRS])


@pytest.mark.benchmark(group="edit_distance_short")
def test_jaro_winkler_short(benchmark):
    benchmark(lambda: [jaro_winkler(a, b) for a, b in SHORT_PAIRS])


@pytest.mark.benchmark(group="edit_distance_short")
def test_sorensen_dice_short(benchmark):
    benchmark(lambda: [sorensen_dice(a, b) for a, b in SHORT_PAIRS])


# --- Hamming on 1000-char strings ---


@pytest.mark.benchmark(group="hamming")
def test_hamming_1000(benchmark):
    pairs = [
        (
            "".join(chr(65 + (i + j) % 26) for j in range(1000)),
            "".join(chr(65 + (i + j + 1) % 26) for j in range(1000)),
        )
        for i in range(6)
    ]
    benchmark(lambda: [hamming(a, b) for a, b in pairs])


# --- Edit distance: paragraph-length strings from War and Peace ---


@pytest.mark.benchmark(group="edit_distance_paragraphs")
def test_levenshtein_paragraphs(benchmark, war_and_peace_paragraphs):
    paras = war_and_peace_paragraphs[:20]
    if len(paras) < 2:
        pytest.skip("not enough paragraphs")
    pairs = list(itertools.pairwise(paras))
    benchmark(lambda: [levenshtein(a, b) for a, b in pairs])


@pytest.mark.benchmark(group="edit_distance_paragraphs")
def test_jaro_winkler_paragraphs(benchmark, war_and_peace_paragraphs):
    paras = war_and_peace_paragraphs[:20]
    if len(paras) < 2:
        pytest.skip("not enough paragraphs")
    pairs = list(itertools.pairwise(paras))
    benchmark(lambda: [jaro_winkler(a, b) for a, b in pairs])


@pytest.mark.benchmark(group="edit_distance_paragraphs")
def test_sorensen_dice_paragraphs(benchmark, war_and_peace_paragraphs):
    paras = war_and_peace_paragraphs[:20]
    if len(paras) < 2:
        pytest.skip("not enough paragraphs")
    pairs = list(itertools.pairwise(paras))
    benchmark(lambda: [sorensen_dice(a, b) for a, b in pairs])


# --- N-gram similarity ---


@pytest.mark.benchmark(group="ngram")
@pytest.mark.parametrize("n", [2, 3, 4], ids=lambda n: f"n={n}")
def test_ngram_jaccard(benchmark, n):
    benchmark(lambda: [ngram_jaccard(a, b, n=n) for a, b in SHORT_PAIRS])


@pytest.mark.benchmark(group="ngram")
@pytest.mark.parametrize("n", [2, 3, 4], ids=lambda n: f"n={n}")
def test_ngram_cosine(benchmark, n):
    benchmark(lambda: [ngram_cosine(a, b, n=n) for a, b in SHORT_PAIRS])


# --- Phonetic ---


@pytest.mark.benchmark(group="phonetic")
def test_soundex_encode(benchmark):
    names = ["Robert", "Rupert", "William", "Williams", "Philadelphia", "Mississippi"]
    benchmark(lambda: [soundex_encode(n) for n in names])


@pytest.mark.benchmark(group="phonetic")
def test_metaphone_encode(benchmark):
    names = ["Robert", "Rupert", "William", "Williams", "Philadelphia", "Mississippi"]
    benchmark(lambda: [metaphone_encode(n) for n in names])


# --- LCS ---


@pytest.mark.benchmark(group="lcs")
def test_lcs_short(benchmark):
    benchmark(lambda: [lcs_distance(a, b) for a, b in SHORT_PAIRS])


@pytest.mark.benchmark(group="lcs")
def test_lcs_100_chars(benchmark):
    a = "abcdefghij" * 10
    b = "abcxefghij" * 10
    benchmark(lambda: lcs_length(a, b))


@pytest.mark.benchmark(group="lcs")
def test_longest_common_substring_100(benchmark):
    a = "abcdefghij" * 10
    b = "abcxefghij" * 10
    benchmark(lambda: longest_common_substring_length(a, b))


# --- Token n-gram similarity: short sentences ---

SENTENCE_PAIRS = [
    ("the quick brown fox", "a quick brown dog"),
    ("New York City is great", "New York City is wonderful"),
    ("machine learning algorithms", "deep learning algorithms"),
    ("United States of America", "United Kingdom of Great Britain"),
    ("natural language processing", "natural language understanding"),
    ("the cat sat on the mat", "the dog sat on the log"),
]


@pytest.mark.benchmark(group="token_ngram_short")
def test_token_jaccard_short(benchmark):
    benchmark(lambda: [token_jaccard(a, b, lowercase=True) for a, b in SENTENCE_PAIRS])


@pytest.mark.benchmark(group="token_ngram_short")
@pytest.mark.parametrize("n", [2, 3], ids=lambda n: f"n={n}")
def test_token_ngram_jaccard_short(benchmark, n):
    benchmark(lambda: [token_ngram_jaccard(a, b, n=n, lowercase=True) for a, b in SENTENCE_PAIRS])


@pytest.mark.benchmark(group="token_ngram_short")
@pytest.mark.parametrize("n", [2, 3], ids=lambda n: f"n={n}")
def test_token_ngram_cosine_short(benchmark, n):
    benchmark(lambda: [token_ngram_cosine(a, b, n=n, lowercase=True) for a, b in SENTENCE_PAIRS])


@pytest.mark.benchmark(group="token_ngram_short")
@pytest.mark.parametrize("n", [2, 3], ids=lambda n: f"n={n}")
def test_token_ngram_overlap_short(benchmark, n):
    benchmark(lambda: [token_ngram_overlap(a, b, n=n, lowercase=True) for a, b in SENTENCE_PAIRS])


# --- Token n-gram similarity: paragraphs from War and Peace ---


@pytest.mark.benchmark(group="token_ngram_paragraphs")
def test_token_jaccard_paragraphs(benchmark, war_and_peace_paragraphs):
    paras = war_and_peace_paragraphs[:20]
    if len(paras) < 2:
        pytest.skip("not enough paragraphs")
    pairs = list(itertools.pairwise(paras))
    benchmark(lambda: [token_jaccard(a, b, lowercase=True) for a, b in pairs])


@pytest.mark.benchmark(group="token_ngram_paragraphs")
def test_token_ngram_jaccard_paragraphs(benchmark, war_and_peace_paragraphs):
    paras = war_and_peace_paragraphs[:20]
    if len(paras) < 2:
        pytest.skip("not enough paragraphs")
    pairs = list(itertools.pairwise(paras))
    benchmark(lambda: [token_ngram_jaccard(a, b, n=2, lowercase=True) for a, b in pairs])


@pytest.mark.benchmark(group="token_ngram_paragraphs")
def test_token_ngram_cosine_paragraphs(benchmark, war_and_peace_paragraphs):
    paras = war_and_peace_paragraphs[:20]
    if len(paras) < 2:
        pytest.skip("not enough paragraphs")
    pairs = list(itertools.pairwise(paras))
    benchmark(lambda: [token_ngram_cosine(a, b, n=2, lowercase=True) for a, b in pairs])

File: benches/bench_algorithms.rs
//! Benchmarks for string distance and similarity algorithms.

use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
use std::hint::black_box;

use _rust::core::algorithms::{
    edit::{DamerauLevenshtein, Hamming, Jaro, JaroWinkler, Levenshtein, Osa, SorensenDice},
    ngram::{
        NgramCosine, NgramJaccard, TokenJaccard, TokenNgramCosine, TokenNgramJaccard,
        TokenNgramOverlap,
    },
    phonetic::{Metaphone, Soundex},
    sequence::Lcs,
    traits::StringDistance,
};

/// Short string pairs for name-matching benchmarks.
const SHORT_PAIRS: &[(&str, &str)] = &[
    ("kitten", "sitting"),
    ("martha", "marhta"),
    ("Robert", "Rupert"),
    ("William", "Williams"),
    ("Philadelphia", "Philadlephia"),
    ("algorithm", "altruistic"),
];

/// Longer strings extracted from War and Peace (first ~500 chars of paragraphs).
fn load_war_and_peace_paragraphs() -> Vec<String> {
    let text = std::fs::read_to_string("tests/fixtures/war_and_peace.txt").unwrap_or_else(|_| {
        // Fallback if file not present
        "The quick brown fox jumps over the lazy dog. ".repeat(100)
    });
    text.split("\n\n")
        .filter(|p| p.len() > 200)
        .take(20)
        .map(|p| p.chars().take(500).collect())
        .collect()
}

fn bench_edit_distance_short(c: &mut Criterion) {
    let mut group = c.benchmark_group("edit_distance/short");

    let algos: Vec<(&str, Box<dyn StringDistance>)> = vec![
        (
            "levenshtein",
            Box::new(Levenshtein) as Box<dyn StringDistance>,
        ),
        ("damerau_levenshtein", Box::new(DamerauLevenshtein)),
        ("osa", Box::new(Osa)),
        ("jaro", Box::new(Jaro)),
        ("jaro_winkler", Box::new(JaroWinkler::default())),
        ("sorensen_dice", Box::new(SorensenDice)),
    ];
    for (name, algo) in &algos {
        group.bench_function(*name, |b| {
            b.iter(|| {
                for (a, s) in SHORT_PAIRS {
                    black_box(algo.distance(a, s).unwrap());
                }
            })
        });
    }
    group.finish();
}

fn bench_hamming(c: &mut Criterion) {
    let h = Hamming;
    let pairs: Vec<(String, String)> = (0..6)
        .map(|i| {
            let a: String = (0..1000)
                .map(|j| ((65 + (i + j) % 26) as u8) as char)
                .collect();
            let b: String = (0..1000)
                .map(|j| ((65 + (i + j + 1) % 26) as u8) as char)
                .collect();
            (a, b)
        })
        .collect();

    c.bench_function("hamming/1000_chars", |b| {
        b.iter(|| {
            for (a, s) in &pairs {
                black_box(h.distance(a, s).unwrap());
            }
        })
    });
}

fn bench_edit_distance_paragraphs(c: &mut Criterion) {
    let paras = load_war_and_peace_paragraphs();
    if paras.len() < 2 {
        return;
    }

    let mut group = c.benchmark_group("edit_distance/paragraphs");
    group.sample_size(10);

    // Compare consecutive paragraph pairs
    let pairs: Vec<(&str, &str)> = paras
        .windows(2)
        .map(|w| (w[0].as_str(), w[1].as_str()))
        .collect();

    let algos: Vec<(&str, Box<dyn StringDistance>)> = vec![
        (
            "levenshtein",
            Box::new(Levenshtein) as Box<dyn StringDistance>,
        ),
        ("jaro_winkler", Box::new(JaroWinkler::default())),
        ("sorensen_dice", Box::new(SorensenDice)),
    ];
    for (name, algo) in &algos {
        group.bench_function(*name, |b| {
            b.iter(|| {
                for (a, s) in &pairs {
                    black_box(algo.distance(a, s).unwrap());
                }
            })
        });
    }
    group.finish();
}

fn bench_ngram_similarity(c: &mut Criterion) {
    let mut group = c.benchmark_group("ngram_similarity");

    for n in [2, 3, 4] {
        let jac = NgramJaccard { n };
        let cos = NgramCosine { n };

        group.bench_with_input(BenchmarkId::new("jaccard", n), &n, |b, _| {
            b.iter(|| {
                for (a, s) in SHORT_PAIRS {
                    black_box(jac.distance(a, s).unwrap());
                }
            })
        });

        group.bench_with_input(BenchmarkId::new("cosine", n), &n, |b, _| {
            b.iter(|| {
                for (a, s) in SHORT_PAIRS {
                    black_box(cos.distance(a, s).unwrap());
                }
            })
        });
    }
    group.finish();
}

fn bench_phonetic(c: &mut Criterion) {
    let mut group = c.benchmark_group("phonetic");
    let soundex = Soundex;
    let metaphone = Metaphone;

    let names = [
        "Robert",
        "Rupert",
        "William",
        "Williams",
        "Philadelphia",
        "Mississippi",
    ];

    group.bench_function("soundex_encode", |b| {
        b.iter(|| {
            for name in &names {
                black_box(soundex.encode(name));
            }
        })
    });

    group.bench_function("metaphone_encode", |b| {
        b.iter(|| {
            for name in &names {
                black_box(metaphone.encode(name));
            }
        })
    });

    group.finish();
}

fn bench_lcs(c: &mut Criterion) {
    let lcs = Lcs;

    c.bench_function("lcs/short", |b| {
        b.iter(|| {
            for (a, s) in SHORT_PAIRS {
                black_box(lcs.lcs_length(a, s));
            }
        })
    });

    // LCS on longer strings (100 chars)
    let a: String = "abcdefghij".repeat(10);
    let s: String = "abcxefghij".repeat(10);
    c.bench_function("lcs/100_chars", |b| {
        b.iter(|| {
            black_box(lcs.lcs_length(&a, &s));
        })
    });
}

/// Sentence pairs for token n-gram benchmarks.
const SENTENCE_PAIRS: &[(&str, &str)] = &[
    ("the quick brown fox", "a quick brown dog"),
    ("New York City is great", "New York City is wonderful"),
    ("machine learning algorithms", "deep learning algorithms"),
    (
        "United States of America",
        "United Kingdom of Great Britain",
    ),
    (
        "natural language processing",
        "natural language understanding",
    ),
    ("the cat sat on the mat", "the dog sat on the log"),
];

fn bench_token_ngram_short(c: &mut Criterion) {
    let mut group = c.benchmark_group("token_ngram/short");

    let tj = TokenJaccard { lowercase: true };
    group.bench_function("token_jaccard", |b| {
        b.iter(|| {
            for (a, s) in SENTENCE_PAIRS {
                black_box(tj.distance(a, s).unwrap());
            }
        })
    });

    for n in [2, 3] {
        let tnj = TokenNgramJaccard { n, lowercase: true };
        let tnc = TokenNgramCosine { n, lowercase: true };
        let tno = TokenNgramOverlap { n, lowercase: true };

        group.bench_with_input(BenchmarkId::new("jaccard", n), &n, |b, _| {
            b.iter(|| {
                for (a, s) in SENTENCE_PAIRS {
                    black_box(tnj.distance(a, s).unwrap());
                }
            })
        });

        group.bench_with_input(BenchmarkId::new("cosine", n), &n, |b, _| {
            b.iter(|| {
                for (a, s) in SENTENCE_PAIRS {
                    black_box(tnc.distance(a, s).unwrap());
                }
            })
        });

        group.bench_with_input(BenchmarkId::new("overlap", n), &n, |b, _| {
            b.iter(|| {
                for (a, s) in SENTENCE_PAIRS {
                    black_box(tno.distance(a, s).unwrap());
                }
            })
        });
    }
    group.finish();
}

fn bench_token_ngram_paragraphs(c: &mut Criterion) {
    let paras = load_war_and_peace_paragraphs();
    if paras.len() < 2 {
        return;
    }

    let mut group = c.benchmark_group("token_ngram/paragraphs");
    group.sample_size(10);

    let pairs: Vec<(&str, &str)> = paras
        .windows(2)
        .map(|w| (w[0].as_str(), w[1].as_str()))
        .collect();

    let tj = TokenJaccard { lowercase: true };
    group.bench_function("token_jaccard", |b| {
        b.iter(|| {
            for (a, s) in &pairs {
                black_box(tj.distance(a, s).unwrap());
            }
        })
    });

    let tnj = TokenNgramJaccard {
        n: 2,
        lowercase: true,
    };
    group.bench_function("bigram_jaccard", |b| {
        b.iter(|| {
            for (a, s) in &pairs {
                black_box(tnj.distance(a, s).unwrap());
            }
        })
    });

    let tnc = TokenNgramCosine {
        n: 2,
        lowercase: true,
    };
    group.bench_function("bigram_cosine", |b| {
        b.iter(|| {
            for (a, s) in &pairs {
                black_box(tnc.distance(a, s).unwrap());
            }
        })
    });

    group.finish();
}

criterion_group!(
    benches,
    bench_edit_distance_short,
    bench_hamming,
    bench_edit_distance_paragraphs,
    bench_ngram_similarity,
    bench_phonetic,
    bench_lcs,
    bench_token_ngram_short,
    bench_token_ngram_paragraphs,
);
criterion_main!(benches);
